Add basic debug tools (plots)

amoodie opened the pull request:
Add some basic debugging plots.

These are helpful for exploring where weird index errors and status bugs arise, with the help of the Python Debugging console.

Add tests, and documentation for the debugging tools.

Example:

```python
self.show_attribute('cell_type')
self.show_ind([15, 16, 17, 19, 20])
plt.show()
```
![image](https://user-images.githubusercontent.com/8801322/87062217-6204dc00-c1d2-11ea-9ba2-0ccd36ab2c85.png)

Commit:
add basic debugging tools

Commit:
add tests, documentation, baseline figures, update github actions workflow, improve code quality, add if to upload when docs build fails

elbeejay commented on pyDeltaRCM/debug_tools.py:
```diff
@@ -0,0 +1,154 @@
+        ax.imshow(getattr(self, attribute),
+                  cmap=plt.get_cmap('viridis'),
+                  interpolation='none')
+        ax.autoscale(False)
```
Can the colorbar be added here? Would be helpful to know when looking at the cell type plot for example, which cell types correspond to which colors on the plot.

elbeejay commented on pyDeltaRCM/debug_tools.py:
```diff
@@ -0,0 +1,154 @@
+
+    Add ``breakpoint()`` to any line in the code to enter the debugger and use
+    these tools interactively at that breakpoint. Note, for Python < 3.7 use
+    pdb.set_trace().
```
Want to stylize it as ``pdb.set_trace()`` to be consistent with ``breakpoint()``?

elbeejay reviewed (approved):
Pretty neat idea. This works for me - as a set of debugging tools I am not concerned about errors or funny plots when you try to `show_attribute` for unusual attributes (like 'seed' which gives an error, or 'iwalk' which sticks a the little iwalk 3x3 in the corner of the grid plot). 

Should be helpful for all future debugging efforts! :+1:

Commit:
size grid to attribute shape

elbeejay commented:
Maybe this is wishful thinking, but do you know if there is a way to pass the command `import matplotlib.pyplot as plt` to the debugger? Unfortunate that the rest of these tools work but you can't display the plot without typing out that line in the debugger itself.

Commit:
add basic type checking

Commit:
add colorbar to show_attribute, update tests and figs

amoodie commented:
Fixed e.g., `iwalk` by making the routine determine the grid size from the attribute itself.
![image](https://user-images.githubusercontent.com/8801322/87188668-e039af00-c2b4-11ea-8682-ac3e83f268d8.png)

Added a colorbar to all plots:
![image](https://user-images.githubusercontent.com/8801322/87188761-00696e00-c2b5-11ea-965a-f1ba9709c8f5.png)

This one is kind of weird, since a lot of these plots are categorical...I think you could do some educated guessing to make the colormaps categorical for certain attributes, but I figured I'd just leave it as "dumb"...

still working on a few things, and why the tests are failing...

Commit:
stylize doc phrase

Commit:
fix path to files

fix broken axis calls

reduce and fix dpi, remove_text

declare figure size, add multi axis test

cleanup code quality

amoodie commented:
> do you know if there is a way to pass the command `import matplotlib.pyplot as plt` to the debugger? Unfortunate that the rest of these tools work but you can't display the plot without typing out that line in the debugger itself.

This is a good idea, but I could not find any way to do it. One option would be to add the import to the top of every module file. This is "bad practice" according to pep etc, but probably fine because we all know that `plt` is sort-of reserved for matplotlib. I'd be fine with adding this though. We could consider it in the future?

Otherwise, I got the tests to pass by removing the text from plots for testing (Windows has a different Freetype library used by matplotlib than *nix platforms. Also, we now explicitly create figures with specified size in the test suite --- for some reason a few figures were being created as different sizes on ubuntu without fixing the size.

Good to merge this, @elbeejay ?
